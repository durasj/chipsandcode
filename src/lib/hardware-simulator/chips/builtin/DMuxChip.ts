import type Chip from '../Chip';
import type { ChipPin } from '../Chip';

/**
 * DMux chip implemented natively
 */
class DMuxChip implements Chip {
  public readonly name = 'DMux' as string;

  private in: boolean[];
  private sel: boolean[];
  private a: boolean[];
  private b: boolean[];

  constructor(private width = 1) {
    this.in = new Array(width).fill(false);
    this.sel = [false];
    this.a = new Array(width).fill(false);
    this.b = new Array(width).fill(false);
  }

  setInput(name: 'in' | 'sel', value: boolean[]) {
    this[name] = value;
  }

  getOutput(name: 'a' | 'b') {
    return this[name];
  }

  getPins() {
    // Constructs map on the fly
    // Pin inspection is less performance sensitive than actual execution

    return new Map([
      ['in', { type: 'output', width: this.width, state: this.in, connections: [] }],
      ['sel', { type: 'input', width: 1, state: this.sel, connections: [] }],
      ['a', { type: 'input', width: this.width, state: this.a, connections: [] }],
      ['b', { type: 'input', width: this.width, state: this.b, connections: [] }],
    ]) satisfies Map<string, ChipPin>;
  }

  public run() {
    const selPin = this.sel[0] ? 'b' : 'a';
    for (const pin of ['a', 'b'] as const) {
      this[pin] = selPin === pin ? this.in : new Array(this.width).fill(false);
    }
  }
}

export default DMuxChip;
