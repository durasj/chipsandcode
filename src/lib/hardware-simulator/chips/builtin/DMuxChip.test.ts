import { describe, expect, it } from 'vitest';

import DMuxChip from './DMuxChip';

describe('DMux Chip', () => {
  it('Provides interface info', async () => {
    const chip = new DMuxChip();

    expect(chip.name).toBe('DMux');
    expect(chip.getPins()).toMatchSnapshot();
  });

  it('Has correct logic', async () => {
    const chip = new DMuxChip();

    const values = [
      [[false], [false], [false], [false]],
      [[true], [false], [true], [false]],
      [[false], [false], [false], [false]],
      [[true], [false], [true], [false]],
      [[false], [true], [false], [false]],
      [[true], [true], [false], [true]],
      [[false], [true], [false], [false]],
      [[true], [true], [false], [true]],
    ];

    values.forEach(([input, sel, a, b]) => {
      chip.setInput('in', input);
      chip.setInput('sel', sel);
      chip.run();

      expect(chip.getOutput('a')).toStrictEqual(a);
      expect(chip.getOutput('b')).toStrictEqual(b);
    });
  });
});
