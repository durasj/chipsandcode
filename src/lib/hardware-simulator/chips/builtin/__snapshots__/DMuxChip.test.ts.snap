// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DMux Chip > Provides interface info 1`] = `
Map {
  "in" => {
    "connections": [],
    "state": [
      false,
    ],
    "type": "output",
    "width": 1,
  },
  "sel" => {
    "connections": [],
    "state": [
      false,
    ],
    "type": "input",
    "width": 1,
  },
  "a" => {
    "connections": [],
    "state": [
      false,
    ],
    "type": "input",
    "width": 1,
  },
  "b" => {
    "connections": [],
    "state": [
      false,
    ],
    "type": "input",
    "width": 1,
  },
}
`;
